### DISCUSSION TITLE
How to exit `Deno.serve` from a signal listener properly?

### DISCUSSION THREAD

**Discussion by AFCMS** [denoland/deno#26165] [Q&A]
I am trying to build a service which is composed of multiple components, with a Deno.serve base web server.

I want to have a zero exit code and I want to handle SIGINT from a signal listener (Deno.addSignalListener) to call the methods closing the different components. While other components work fine with this setup, Deno.serve seems to handle the signal himself and return a 130 code:

```ts
Deno.addSignalListener("SIGINT", async () => {
  console.log("Caught SIGINT, shutting down...");
  console.log("Shutting down server...");
  await server.shutdown();
  console.log("Shutting down client...");
  await client.destroy();
  await new Promise((resolve) => setTimeout(resolve, 1000));
  db.close();
  kv.close();
  console.log("Exiting...");
  Deno.exit(0);
});
```
It look like this in the console, so basically returning the error code (that I didn't ask for) and then the output still get printed? :

![image](https://github.com/user-attachments/assets/5170dfc4-0e69-42b0-aafc-18ea6458bda8)

**Comment by Chaitanya-Kyp** [comment]
To properly exit `Deno.serve` with a zero exit code, use an `AbortController` to signal the server shutdown. This prevents Deno from automatically returning a 130 code:

```ts
const controller = new AbortController();
const { signal } = controller;

Deno.serve({ signal }, async (req) => {
  return new Response("Hello World");
});

Deno.addSignalListener("SIGINT", async () => {
  console.log("Caught SIGINT, shutting down...");
  controller.abort(); // Gracefully shut down server
  await client.destroy();
  db.close();
  kv.close();
  Deno.exit(0); // Ensure zero exit code
});
```

Hope it helps!
